src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala

```diff
@@ -89,10 +98,17 @@ class RegularSync(
     // The case that waitingForActor is defined (we are waiting for some response),
     // can happen when we are on top of the chain and currently handling newBlockHashes message
 
-    if (waitingForActor.isEmpty)
-      askForHeaders()
-    else
+    if (waitingForActor.isEmpty) {
```

Comment:
(super minor) is it worthy to replace those nested `if`that check for `isEmpty` with something else like map or pattern matching?

Reply:
What would that look like? TBH I believe this whole class deserves a full rewrite. I have yet to write a single change in it without introducing a bug. There should be clearly defined state (mode of operation), that determines the way we handle all the mutables - currently it's the other way around.